### PR TITLE
Add support for scraping Sofascore

### PR DESCRIPTION
**Pull request by ksbharaj**
Hi All.
I have included some code to scrape Sofascore. 
It's fairly basic now, focussed only on the top 5 leagues. I can add more functions going forward.
Happy to get any feedback to help me improve it.

### CODE REVIEW

**Comment by probberechts** [comment]
Hi. Thanks for your contribution. Does Sofascore provide any data that is not yet covered by the scrapers that are already implemented? I am somewhat hesitant to add more sources that provide exactly the same data, because each data source comes with a maintenance cost.

**Comment by ksbharaj** [comment]
Apologies for my delayed response. Github's stopped emailing me notifications - my responses will be quicker next time.
Yes, Sofascore does. I've listed its distinguishing qualities below:
- Inclusion of match momentum data
- Inclusion of player heatmap and average positions (possible to extract from WhoScored event data, but that requires far more code and takes much, much more time to scrape than Sofascore does)
- Inclusion of all useful details in ShotMap (such as xG, xGOT, shot time and location). The other sources tend to lack at least one of these in their shotmaps.

I understand your concerns on the maintenance side. However, I'm happy to manage it for SofaScore going forward if possible.

**Comment by ksbharaj** [comment]
Also worth mentioning is that I haven't included the listed features above in this initial commit- I intended for the initial commit to be simple. I can add it later on with relative ease.

**Comment by probberechts** [comment]
Thanks for the clarification. For me, the most important requirement is that a few tests are added and that the docs are updated. Extra features can be added later.

**Comment by ksbharaj** [comment]
I agree. I will work towards that then. Thank you!

**Comment by ksbharaj** [comment]
Hi. Please let me know if there's anything else required to merge this PR. 
I'll look to get the /docs updated in the next couple of days too. Thanks!

File: soccerdata/sofascore.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+SOFASCORE_DATADIR = DATA_DIR / 'Sofascore'
+SOFASCORE_API = 'https://api.sofascore.com/api/v1/'
+
+BIG_FIVE_DICT = {
+    "Serie A": "ITA-Serie A",
+    "Ligue 1": "FRA-Ligue 1",
+    "LaLiga": "ESP-La Liga",
+    "Premier League": "ENG-Premier League",
+    "Bundesliga": "GER-Bundesliga",
+}
```
Comment (probberechts) [inline soccerdata/sofascore.py:23]: This was copy-pasted from the FBref scraper. I assume it is not needed for Sofascore.

File: soccerdata/sofascore.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        leagues = []
+        for k in data['uniqueTournaments']:
+            if k['name'] in BIG_FIVE_DICT:
+                leagues.append(
+                    {
+                        'region': BIG_FIVE_DICT[k['name']].split("-")[0],
+                        'league_id': k['id'],
+                        'league': BIG_FIVE_DICT[k['name']],
+                    }
+                )
```
Comment (probberechts) [inline soccerdata/sofascore.py:109]: You should use the `_translate_league` helper instead.

File: soccerdata/sofascore.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+            print
+            url = SOFASCORE_API + 'unique-tournament/{}/seasons'
+            filepath = self.data_dir / filemask.format(lkey)
+            reader = self.get(url.format(league.league_id), filepath)
+            data = json.load(reader)['seasons']
+            for season in data:
+                seasons.append(
+                    {
+                        'league': lkey,
+                        'season': season['year'].replace("/", ""),
```
Comment (probberechts) [inline soccerdata/sofascore.py:133]: Use the `season_code` helper to convert seasons. This ensures that season codes are consistent everywhere.

File: soccerdata/sofascore.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            'away_score',
+            'game_id',
+        ]
+
+        df_seasons = self.read_seasons()
+        all_schedules = []
+        for (lkey, skey), season in df_seasons.iterrows():
+            # filepath = self.data_dir / filemask.format(lkey, skey)
+            url1 = urlmask1.format(season['league_id'], season['season_id'])
+            reader1 = self.get(url1)
```
Comment (probberechts) [inline soccerdata/sofascore.py:220]: Why do you not use caching here?

File: soccerdata/sofascore.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+            season_data = json.load(reader1)
+            rounds = season_data['rounds']
+
+            for round in rounds:
+                url2 = urlmask2.format(season['league_id'], season['season_id'], round['round'])
+                reader2 = self.get(url2)
```
Comment (probberechts) [inline soccerdata/sofascore.py:226]: Why do you not use caching here?

File: soccerdata/sofascore.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+                        all_schedules.append(
+                            {
+                                'league': lkey,
+                                'season': skey,
+                                'round': "None",
+                                'week': _match['roundInfo']['round'],
+                                'date': datetime.datetime.fromtimestamp(
+                                    _match['startTimestamp']
+                                ).strftime("%Y-%m-%d %H:%M:%S%z"),
```
Comment (probberechts) [inline soccerdata/sofascore.py:245]: Remove the strftime method here and drop the conversion to datetime on line 262.

File: tests/conftest.py [hunk 7]
```diff
@@ -69,3 +69,15 @@ def understat_epl_1516() -> sd.Understat:
 def understat_epl_9091() -> sd.Understat:
     """Return a correctly initialized instance of Understat filtered by league: Premier League."""
     return sd.Understat("ENG-Premier League", "90-91")
+
+
+@pytest.fixture
+def sofascore_epl_1516() -> sd.Sofascore:
+    """Return a correctly initialized instance of Understat filtered by league: Premier League."""
```
Comment (probberechts) [inline tests/conftest.py:76]: Understat --> Sofascore

File: tests/conftest.py [hunk 8]
```diff
@@ -69,3 +69,15 @@ def understat_epl_1516() -> sd.Understat:
...
+    return sd.Sofascore("ENG-Premier League", "15-16")
+
+
+@pytest.fixture
+def sofascore_epl_9091() -> sd.Sofascore:
+    """Return a correctly initialized instance of Understat filtered by league: Premier League."""
```
Comment (probberechts) [inline tests/conftest.py:82]: Understat --> Sofascore

**Review by probberechts — CHANGES_REQUESTED** [review]
This starts to look good! I've added a few comments and noticed that you still have to update the following files:
- `docs/datasources/index.rst`
- `docs/reference/index.rst`
- `docs/howto/custom-leagues.rst`

**Comment by ksbharaj** [comment]
Thank you for reviewing! I'll look to get these addressed asap (next couple of days at most)

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -52,6 +52,6 @@ repos:
         types: [text]
         stages: [commit, push, manual]
   - repo: https://github.com/pre-commit/mirrors-prettier
-    rev: v2.4.1
+    rev: v4.0.0-alpha.8
```
[the remark this answers is not in the document]
Comment (ksbharaj) [reply .pre-commit-config.yaml]: I have changed the version to v3.1.0 (not an alpha release). Let me know if that's OK

File: soccerdata/sofascore.py (hunk 1 above)
Comment (ksbharaj) [reply soccerdata/sofascore.py]: Yes, you are correct. I have excluded this now

File: soccerdata/sofascore.py (hunk 2 above)
Comment (ksbharaj) [reply soccerdata/sofascore.py]: Included now

File: soccerdata/sofascore.py (hunk 3 above)
Comment (ksbharaj) [reply soccerdata/sofascore.py]: Included now

File: soccerdata/sofascore.py (hunk 4 above)
Comment (ksbharaj) [reply soccerdata/sofascore.py]: I have implemented caching now. Apologies for that

File: soccerdata/sofascore.py (hunk 5 above)
Comment (ksbharaj) [reply soccerdata/sofascore.py]: I have implemented caching now. Apologies for that

File: soccerdata/sofascore.py (hunk 6 above)
Comment (ksbharaj) [reply soccerdata/sofascore.py]: Modified as requested

File: tests/conftest.py (hunk 7 above)
Comment (ksbharaj) [reply tests/conftest.py]: Apologies for the error. Now resolved

File: tests/conftest.py (hunk 8 above)
Comment (ksbharaj) [reply tests/conftest.py]: Apologies for the error. Now resolved

File: tests/conftest.py [hunk 10]
```diff
@@ -69,3 +69,15 @@ def understat_epl_1516() -> sd.Understat:
...
+
+
+@pytest.fixture
+def sofascore_epl_1516() -> sd.Sofascore:
+    """Return a correctly initialized instance of Understat filtered by league: Premier League."""
+    return sd.Sofascore("ENG-Premier League", "15-16")
+
+
+@pytest.fixture
+def sofascore_epl_9091() -> sd.Sofascore:
```
[the remark this answers is not in the document]
Comment (ksbharaj) [reply tests/conftest.py]: I had copied the test used for Understat. 
I have removed the fixture test now, but would you recommend I modify it instead or include other (similar) tests?

[the remark this answers is not in the document]
Comment (probberechts) [reply tests/conftest.py]: It seems like you removed the fixture but not the test case. What I meant is that I would put the fixture inside the `test_sofscore.py` file or create the scraper instance inside the `test_read_seasons_empty` test case.

**Comment by probberechts** [comment]
Could you fix the merge conflicts, please? The tests currently fail due to an issue with pipx installs in GHA but I've just fixed this in the master branch.

File: tests/conftest.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ksbharaj) [reply tests/conftest.py]: Thanks- sorry for that. I have made the change now. I've created a scraper instance inside the `test_read_seasons_empty` test case in 'test_sofascore.py'

**Comment by tymsoncyferki** [comment]
hey @ksbharaj, are you planning to add more detailed methods such as reading match stats and player stats? cheers

**Comment by ksbharaj** [comment]
> hey @ksbharaj, are you planning to add more detailed methods such as reading match stats and player stats? cheers

Hey! Sorry been busy with other projects so haven't had chance to work on this but I will do so before the end of this month!

**Comment by tymsoncyferki** [comment]
> Hey! Sorry been busy with other projects so haven't had chance to work on this but I will do so before the end of this month!

Glad to hear it! I see that you are not necessarily scraping it but using the API. Does it mean that all the methods will be available only for the leagues present in [this](https://api.sofascore.com/api/v1/config/unique-tournaments/EN/football) JSON?

**Comment by ksbharaj** [comment]
I'll need to start scraping because sofascore have changed things a bit since I last updated. 
I think it will be for those only- were you looking for any others?

**Comment by tymsoncyferki** [comment]
Yeah, I was hoping it would be possible to scrape any league present on Sofascore (I am particularly interested in [second Polish division](https://www.sofascore.com/tournament/football/poland/i-liga/229#id:61452)). But you know, do your thing, and if it will not work on this league I will just create my own scraper;)

**Pull request merged**